File: Dynamic_Programming/915.Partition_Array_into_Disjoint_Intervals/Python/915.Partition_Array_into_Disjoint_Intervals.py
class Solution(object):
    def partitionDisjoint(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #============================================#
        # Dynamic-programming based traversal method #
        #============================================#
        if (len(nums) == 2): #Issue/Boundary-case handling
            return 1

        ############
        #Initialize
        ##### Length of nums array #####
        len_nums = len(nums)

        ##### Record dynamic-programming array (i.e. prefix, suffix) #####
        record_prefix_dp_arry, record_suffix_dp_arry = [0 for _ in range(len_nums)], [0 for _ in range(len_nums)]

        record_prefix_dp_arry[0] = (-1)
        record_prefix_dp_arry[1] = nums[0]

        record_suffix_dp_arry[(len_nums - 1)] = (1e6 + 1)
        record_suffix_dp_arry[(len_nums - 2)] = nums[(len_nums - 1)]


        ##########################################
        #Dynamic-programming based loop traversal
        ##### Step 1: Record indexed-value with dynamic-programming array (i.e. prefix, suffix) #####
        for nums_idx in range(2, len_nums): #Prefix
            record_prefix_dp_arry[nums_idx] = max(record_prefix_dp_arry[(nums_idx - 1)], nums[(nums_idx - 1)]) #Keep updating/overwriting

        for nums_idx in range((len_nums - 3), (-1), (-1)): #Suffix
            record_suffix_dp_arry[nums_idx] = min(record_suffix_dp_arry[(nums_idx + 1)], nums[(nums_idx + 1)]) #Keep updating/overwriting

        ##### Step 2: Looped-traversal with recorded dynamic-programming array (i.e. prefix, suffix) #####
        for record_idx in range(1, len_nums):

            ##### Check if the current indexed-value matched conditions or not #####
            if ((nums[record_idx] >= record_prefix_dp_arry[record_idx]) and (record_suffix_dp_arry[record_idx] >= record_prefix_dp_arry[record_idx])):
                return record_idx

        return (-1)

File: Dynamic_Programming/915.Partition_Array_into_Disjoint_Intervals/C++/915.Partition_Array_into_Disjoint_Intervals.cpp
class Solution {
public:
    int partitionDisjoint(vector<int> & nums) 
    {
        //====================================//
        // Input type:                        //
        //   - vector<int> ref. (nums vector) //
        // Return type:                       //
        //   - int (result minimun length)    //
        //====================================//
        if (nums.size() == 2) { return 1; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record dynamic-programming vector (i.e. prefix, suffix) /////
        vector<int> record_prefix_dp_vect(len_nums, 0), record_suffix_dp_vect(len_nums, 0);

        record_prefix_dp_vect[0] = (-1);
        record_prefix_dp_vect[1] = nums[0];

        record_suffix_dp_vect[(len_nums - 1)] = (1e6 + 1);
        record_suffix_dp_vect[(len_nums - 2)] = nums[(len_nums - 1)];


        /*Dynamic-programming based loop traversal*/
        ///// Step 1: Record indexed-value with dynamic-programming vector (i.e. prefix, suffix) /////
        for (int nums_idx = 2; (nums_idx < len_nums); (nums_idx++)) //Prefix
        {
            record_prefix_dp_vect[nums_idx] = max(record_prefix_dp_vect[(nums_idx - 1)], nums[(nums_idx - 1)]); //Keep updating, overwriting

        } //Prefix

        for (int nums_idx = (len_nums - 3); (nums_idx >= 0); (nums_idx--)) //Suffix
        {
            record_suffix_dp_vect[nums_idx] = min(record_suffix_dp_vect[(nums_idx + 1)], nums[(nums_idx + 1)]); //Keep updating, overwriting

        } //Suffix

        ///// Step 3: Looped-traversal with recorded dynamic-programming vector (i.e. prefix, suffix) /////
        for (int record_idx = 1; (record_idx < len_nums); (record_idx++)) //Whole
        {
            ///// Check if the current indexed-value matched conditions or not /////
            if ((nums[record_idx] >= record_prefix_dp_vect[record_idx]) && (record_suffix_dp_vect[record_idx] >= record_prefix_dp_vect[record_idx]))
            {
                return record_idx;
            }

        } //Whole

        return (-1);
    }
};
